components/ai_chat/core/browser/conversation_handler_unittest.cc

```diff
@@ -2528,6 +2532,148 @@ TEST_F(ConversationHandlerUnitTest, RateMessage) {
+  auto* tool1_ptr = tool1.get();
+  tool1->set_requires_user_interaction_before_handling(false);
+  std::vector<std::unique_ptr<Tool>> tools;
+  tools.push_back(std::move(tool1));
+  conversation_handler_->SetToolsForTesting(std::move(tools));
+
+  bool tool_response_generation_started = false;
+
+  // We should only add the initial human entry and the first assistant response
+  // but no further human or assistant entries.
+  // Verify it's never called after the tool response generation has started.
+  EXPECT_CALL(observer, OnConversationEntryAdded)
+      .Times(2)
+      .WillRepeatedly(testing::InvokeWithoutArgs(
+          [&]() { EXPECT_FALSE(tool_response_generation_started); }));
+
+  base::RunLoop run_loop;
+  testing::Sequence seq;
+
+  // First call to engine mocks the use tool request
+  EXPECT_CALL(*engine, GenerateAssistantResponse)
+      .InSequence(seq)
+      .WillOnce(testing::DoAll(
+          testing::WithArg<6>(
+              [](EngineConsumer::GenerationDataCallback callback) {
+                callback.Run(EngineConsumer::GenerationResultData(
+                    mojom::ConversationEntryEvent::NewCompletionEvent(
+                        mojom::CompletionEvent::New("Ok, going to check...")),
+                    std::nullopt));
+              }),
+          testing::WithArg<6>(
+              [](EngineConsumer::GenerationDataCallback callback) {
+                callback.Run(EngineConsumer::GenerationResultData(
+                    mojom::ConversationEntryEvent::NewToolUseEvent(
+                        mojom::ToolUseEvent::New("weather_tool", "tool_id_1",
+                                                 "{\"location\":\"New York\"}",
+                                                 std::nullopt)),
+                    std::nullopt));
+              }),
+          testing::WithArg<7>(
+              [&](EngineConsumer::GenerationCompletedCallback callback) {
+                std::move(callback).Run(
+                    base::ok(EngineConsumer::GenerationResultData(
+                        mojom::ConversationEntryEvent::NewCompletionEvent(
+                            mojom::CompletionEvent::New("")),
+                        std::nullopt)));
+              })));
+
+  EXPECT_CALL(*tool1_ptr, UseTool(StrEq("{\"location\":\"New York\"}"), _))
+      .WillOnce(testing::WithArg<1>([&](Tool::UseToolCallback callback) {
+        std::vector<mojom::ContentBlockPtr> result;
+        result.push_back(mojom::ContentBlock::NewTextContentBlock(
+            mojom::TextContentBlock::New("Weather in New York: 72°F")));
+        std::move(callback).Run(std::move(result));
+      }));
+
+  // Mock engine to return no new entry after the tool is used
+  EXPECT_CALL(*engine, GenerateAssistantResponse)
+      .InSequence(seq)
+      .WillOnce(testing::DoAll(
+          testing::InvokeWithoutArgs(
+              [&]() { tool_response_generation_started = true; }),
```

Comment:
could we move this into the `testing::WithArg<7>` callback and get rid of the `testing::DoAll`?

Reply:
Unless I'm misunderstanding and they run one by one, instead of all on the same call to `GenerateAssistantResponse`?